src/ghostwriter/models/v1/mapping.py

```diff
@@ -0,0 +1,180 @@
+            raise HookNotFoundError(f"Hook {hook} could not be loaded")
+        if hook.startswith("ghostwriter.hooks."):
+            hookname = hook[len("ghostwriter.hooks.") :]
+        else:
+            hookname = hook
+        try:
+            obj = getattr(hooks, hookname)
+        except AttributeError as exc:
+            raise HookNotFoundError(
+                f"Hookname {hook} could not be loaded: {exc}"
+            ) from exc
+        retval.append(obj)
+    return retval
+
+
+class MapRule(BaseModel):
+    """Instructions for rewriting a map."""
+
+    source_prefix: Annotated[
+        str,
+        Field(
+            title="Source route",
+            description=(
+                "Generic route prefix to be substituted."
+                " Everything after this prefix will be collected in"
+                " `path` for purposes of substitution in the target"
+                " route."
+            ),
+            examples=["/tutorials/"],
+        ),
+        AfterValidator(canonicalize_source_route),
+    ]
+
+    target: Annotated[
+        str,
+        Field(
+            title="Target route",
+            description=(
+                "Route prefix (with placeholder[|s]) for the resolved"
+                " route. The placeholders are Python string.Template"
+                " names.  ${base_url} and ${user} are determined from"
+                " the request context, and ${path} is everything following"
+                " the source route string in the input route."
+            ),
+            examples=["${base_url}/nb/user/${user}/lab/tree/${path}.ipynb"],
+        ),
+    ]
+
+    hooks: Annotated[
+        None | list[Callable[[Parameters], Awaitable[None]]],
+        Field(
+            title="Pre-substitution hooks",
+            description=(
+                "List of hooks to be called in order before substituting"
+                " and returning the route. Each hook takes as input a"
+                " ~ghostwriter.models.substitution.Parameters object"
+                " and returns `None`. In the event of hook failure, it"
+                " should raise an exception. These hooks should be"
+                " idempotent, which in the usual case, means not"
+                " overwriting an extant file, since the common case is to"
+                " build or download a file in a user context before"
+                " returning the URL pointing to it."
+            ),
+        ),
+        BeforeValidator(resolve_hooks),
+    ] = None
+
+    skip_hooks: Annotated[
+        bool,
+        Field(
+            title=(
+                "Do not run hooks before substituting routes. You should"
+                " only use this for testing."
+            ),
+        ),
+    ] = False
+
+    async def run_hooks(self, params: Parameters) -> None:
+        """Run hooks for a route."""
+        if self.skip_hooks:
+            return
+        if self.hooks is None:
+            return
```

Comment:
```suggestion
        if self.hooks is None or self.skip_hooks:
```

Reply:
I think I'm just going to drop skip_hooks.  I thought I'd need it for testing, but I don't.